api/envoy/config/core/v3/base.proto

```diff
@@ -324,9 +324,18 @@ message HeaderValue {
   // The same :ref:`format specifier <config_access_log_format>` as used for
   // :ref:`HTTP access logging <config_access_log>` applies here, however
   // unknown header values are replaced with the empty string instead of ``-``.
+  // Header value is encoded as string. This does not work for non-utf8 characters.
+  // Only one of ``value`` or ``value_bytes`` can be set.
   string value = 2 [
-    (validate.rules).string = {max_bytes: 16384 well_known_regex: HTTP_HEADER_VALUE strict: false}
+    (validate.rules).string = {max_bytes: 16384 well_known_regex: HTTP_HEADER_VALUE strict: false},
+    (udpa.annotations.field_migrate).oneof_promotion = "value_type"
   ];
+
+  oneof value_type {
```

Comment:
yea, as I mentioned what you need is adding `one_of_promotion` to this new field as well.
```
bytes value_bytes = 3 [(validate.rules).bytes = {min_len: 0 max_len: 16384}];
(udpa.annotations.field_migrate).oneof_promotion = "value_type"
```
Basically this annotations is a reminder to future major API update that coverts those field to to oneof. This is the reason why I also suggested  the configuration validation . Because in current API, those two fields are technically not oneof. They will become oneof in next API promotion (if happens)  

I think that http_connection_manager.proto didn't do it right.

Reply:
[Here](https://github.com/envoyproxy/envoy/blame/61958d40059a3d901b6f0e453d769f58a3dd2e6a/api/envoy/config/route/v3/route_components.proto#LL796C12-L796C26) is a better example: `cluster_header` and `cluster`

Reply:
How about the stats_flush in bootstrap.proto?https://github.com/envoyproxy/envoy/blob/61958d40059a3d901b6f0e453d769f58a3dd2e6a/api/envoy/config/bootstrap/v3/bootstrap.proto#L198

BTW, that cluster_header is also promoted to a oneof type: cluster_specifier here: https://github.com/envoyproxy/envoy/blob/61958d40059a3d901b6f0e453d769f58a3dd2e6a/api/envoy/config/route/v3/route_components.proto#L1019

Reply:
That `cluster_specifier` you referring to is in `RouteAction`. which is not related to the example I shared. In other words, It is not the promoted result of my example. You can also see it from the timeline that it was added 4 years ago while my example is 1 year ago.

IMHO oneof you have here is redundant. It is oneof with only single element.

Reply:
@adisuissa could you please comment on this?  My intuitively interprectation of the text below from the [Style Guide](https://github.com/envoyproxy/envoy/blob/main/api/STYLE.md#field-level) is that the new field is a oneof field, and the existing field will be moved into that oneof in the next major API change.
"
[(udpa.annotations.field_migrate).oneof_promotion = "<oneof name>"] to denote that the field will be promoted to a given oneof in the next API major version.
"

Reply:
There is not going to be another major xDS API major version for a long time (if ever). In the meanwhile, we use the `oneof_promotion` fields to indicate that this is logically a `oneof`, even though it's not enforced in the proto type system. The advice in https://github.com/envoyproxy/envoy/pull/27865/files#r1232707830 will work.

Reply:
I see. Thanks for the explanation.